File: CashuWallet/Core/NIP17.swift
import Foundation

struct NostrIncomingEvent: Codable, Hashable {
    let id: String
    let pubkey: String
    let createdAt: Int64
    let kind: Int
    let tags: [[String]]
    let content: String
    let sig: String

    enum CodingKeys: String, CodingKey {
        case id
        case pubkey
        case createdAt = "created_at"
        case kind
        case tags
        case content
        case sig
    }
}

struct NostrRumor: Hashable {
    let id: String
    let pubkey: String
    let createdAt: Int64
    let kind: Int
    let tags: [[String]]
    let content: String
}

enum NIP17 {
    enum Error: Swift.Error {
        case decryptFailed
        case invalidSeal
        case invalidRumor
        case canonicalizeFailed
    }

    /// Unwrap an incoming kind:1059 gift wrap addressed to us. Returns the inner kind:14 rumor.
    static func unwrap(giftWrap event: NostrIncomingEvent, recipientPrivateKey: Data) throws -> NostrRumor {
        // Layer 1: decrypt wrap content with our private key + wrap.pubkey (ephemeral)
        let sealJSON: String
        do {
            sealJSON = try NIP44.decrypt(
                payload: event.content,
                recipientPrivateKey: recipientPrivateKey,
                senderPubkeyHex: event.pubkey
            )
        } catch {
            throw Error.decryptFailed
        }
        guard let seal = decodeEvent(sealJSON) else { throw Error.invalidSeal }
        guard seal.kind == 13 else { throw Error.invalidSeal }

        // Layer 2: decrypt seal content with our private key + seal.pubkey (real sender)
        let rumorJSON: String
        do {
            rumorJSON = try NIP44.decrypt(
                payload: seal.content,
                recipientPrivateKey: recipientPrivateKey,
                senderPubkeyHex: seal.pubkey
            )
        } catch {
            throw Error.decryptFailed
        }
        guard let rumor = decodeRumor(rumorJSON, expectedAuthor: seal.pubkey) else {
            throw Error.invalidRumor
        }
        return rumor
    }

    // MARK: - JSON helpers

    private static func decodeEvent(_ jsonString: String) -> NostrIncomingEvent? {
        guard let data = jsonString.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(NostrIncomingEvent.self, from: data)
    }

    private static func decodeRumor(_ jsonString: String, expectedAuthor: String) -> NostrRumor? {
        guard let data = jsonString.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        let pubkey = (obj["pubkey"] as? String) ?? expectedAuthor
        guard pubkey == expectedAuthor else { return nil }
        let id = (obj["id"] as? String) ?? ""
        let createdAt = (obj["created_at"] as? NSNumber)?.int64Value
            ?? (obj["created_at"] as? Int64)
            ?? Int64((obj["created_at"] as? Int) ?? 0)
        let kind = (obj["kind"] as? Int) ?? 0
        let tags = (obj["tags"] as? [[String]]) ?? []
        let content = (obj["content"] as? String) ?? ""
        return NostrRumor(id: id, pubkey: pubkey, createdAt: createdAt, kind: kind, tags: tags, content: content)
    }
}

File: android/app/src/main/java/com/cashu/me/Core/NIP17.kt
package com.cashu.me.Core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull

@Serializable
data class NostrIncomingEvent(
    val id: String,
    val pubkey: String,
    @SerialName("created_at")
    val createdAt: Long,
    val kind: Int,
    val tags: List<List<String>>,
    val content: String,
    val sig: String,
)

data class NostrRumor(
    val id: String,
    val pubkey: String,
    val createdAt: Long,
    val kind: Int,
    val tags: List<List<String>>,
    val content: String,
)

internal object NIP17 {
    private val json = Json { ignoreUnknownKeys = true }

    fun unwrap(giftWrap: NostrIncomingEvent, recipientPrivateKey: ByteArray): NostrRumor {
        val sealJson = NIP44.decrypt(
            payload = giftWrap.content,
            recipientPrivateKey = recipientPrivateKey,
            senderPubkeyHex = giftWrap.pubkey,
        )
        val seal = json.decodeFromString<NostrIncomingEvent>(sealJson)
        require(seal.kind == 13) { "Invalid NIP-17 seal." }
        val rumorJson = NIP44.decrypt(
            payload = seal.content,
            recipientPrivateKey = recipientPrivateKey,
            senderPubkeyHex = seal.pubkey,
        )
        return decodeRumor(rumorJson, expectedAuthor = seal.pubkey)
    }

    private fun decodeRumor(rawJson: String, expectedAuthor: String): NostrRumor {
        val fields = json.parseToJsonElement(rawJson).jsonObject
        val pubkey = fields["pubkey"]?.jsonPrimitive?.contentOrNull ?: expectedAuthor
        require(pubkey == expectedAuthor) { "Invalid NIP-17 rumor author." }
        val tags = fields["tags"]?.jsonArray?.map { tag ->
            tag.jsonArray.mapNotNull { value -> value.jsonPrimitive.contentOrNull }
        }.orEmpty()
        return NostrRumor(
            id = fields["id"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            pubkey = pubkey,
            createdAt = fields["created_at"]?.jsonPrimitive?.longOrNull
                ?: fields["created_at"]?.jsonPrimitive?.intOrNull?.toLong()
                ?: 0,
            kind = fields["kind"]?.jsonPrimitive?.intOrNull ?: 0,
            tags = tags,
            content = fields["content"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        )
    }
}
